File: benchmarks/cross-language/java/ThreadRing.java
// Java Thread Ring Benchmark (Savina-style)
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CountDownLatch;

public class ThreadRing {
    private static final int RING_SIZE = 100;
    private static int getMessages() {
        String env = System.getenv("BENCHMARK_MESSAGES");
        return env != null ? Integer.parseInt(env) : 100000;
    }
    private static final int NUM_HOPS = getMessages();

    static class RingNode implements Runnable {
        private final BlockingQueue<Integer> inbox;
        private BlockingQueue<Integer> next;
        private final CountDownLatch done;
        private int received = 0;

        RingNode(CountDownLatch done) {
            // Queue size 8 reduces context switching while preserving ordering
            this.inbox = new ArrayBlockingQueue<>(8);
            this.done = done;
        }

        void setNext(BlockingQueue<Integer> next) {
            this.next = next;
        }

        BlockingQueue<Integer> getInbox() {
            return inbox;
        }

        @Override
        public void run() {
            try {
                while (true) {
                    int token = inbox.take();
                    received++;
                    if (token == 0) {
                        done.countDown();
                        return;
                    }
                    next.put(token - 1);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        int getReceived() {
            return received;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Java Thread Ring Benchmark ===");
        System.out.println("Ring size: " + RING_SIZE + ", Hops: " + NUM_HOPS + "\n");

        CountDownLatch done = new CountDownLatch(1);
        RingNode[] nodes = new RingNode[RING_SIZE];
        Thread[] threads = new Thread[RING_SIZE];

        // Create nodes
        for (int i = 0; i < RING_SIZE; i++) {
            nodes[i] = new RingNode(done);
        }

        // Link nodes in ring
        for (int i = 0; i < RING_SIZE; i++) {
            nodes[i].setNext(nodes[(i + 1) % RING_SIZE].getInbox());
        }

        // Start threads
        for (int i = 0; i < RING_SIZE; i++) {
            threads[i] = new Thread(nodes[i]);
            threads[i].start();
        }

        long start = System.nanoTime();

        // Inject token
        nodes[0].getInbox().put(NUM_HOPS);

        // Wait for completion
        done.await();

        long end = System.nanoTime();
        double elapsedSec = (end - start) / 1e9;

        int totalMessages = NUM_HOPS + 1;
        double throughput = totalMessages / elapsedSec / 1e6;
        double nsPerMsg = elapsedSec * 1e9 / totalMessages;

        System.out.printf("ns/msg:         %.2f%n", nsPerMsg);
        System.out.printf("Throughput:     %.2f M msg/sec%n", throughput);

        // Cleanup - interrupt remaining threads
        for (Thread t : threads) {
            t.interrupt();
        }
    }
}

File: benchmarks/cross-language/scala/thread_ring.scala
// Scala Akka Thread Ring Benchmark (Savina-style)
import akka.actor.{Actor, ActorRef, ActorSystem, Props}
import scala.concurrent.duration._
import scala.concurrent.{Await, Promise}

case class SetNext(next: ActorRef)
case class Token(hops: Int)
case class RingDone(received: Int)

class RingNodeActor(parent: ActorRef) extends Actor {
  var next: ActorRef = null
  var received = 0

  def receive = {
    case SetNext(n) =>
      next = n
    case Token(0) =>
      received += 1
      parent ! RingDone(received)
    case Token(hops) =>
      received += 1
      next ! Token(hops - 1)
  }
}

class ParentActor(promise: Promise[Int], ringSize: Int, numHops: Int) extends Actor {
  var nodes: Array[ActorRef] = null

  override def preStart(): Unit = {
    // Create ring nodes
    nodes = Array.tabulate(ringSize)(i =>
      context.actorOf(Props(classOf[RingNodeActor], self), s"node$i")
    )

    // Link them in a ring
    for (i <- 0 until ringSize) {
      nodes(i) ! SetNext(nodes((i + 1) % ringSize))
    }

    // Inject token
    nodes(0) ! Token(numHops)
  }

  def receive = {
    case RingDone(received) =>
      promise.success(received)
      context.system.terminate()
  }
}

object ThreadRingBenchmark extends App {
  val ringSize = 100
  val numHops = sys.env.get("BENCHMARK_MESSAGES").flatMap(s => scala.util.Try(s.toInt).toOption).getOrElse(100000)
  println("=== Scala Akka Thread Ring Benchmark ===")
  println(s"Ring size: $ringSize, Hops: $numHops\n")

  val system = ActorSystem("ThreadRing")
  val promise = Promise[Int]()

  val startTime = System.nanoTime()

  val parent = system.actorOf(Props(classOf[ParentActor], promise, ringSize, numHops))

  val received = Await.result(promise.future, 120.seconds)
  val endTime = System.nanoTime()

  val elapsed = (endTime - startTime).toDouble / 1e9
  val totalMessages = numHops + 1

  if (received != totalMessages) {
    println(s"VALIDATION FAILED: expected $totalMessages, got $received")
  }

  val throughput = totalMessages / elapsed / 1e6
  val nsPerMsg = elapsed * 1e9 / totalMessages

  println(f"ns/msg:         $nsPerMsg%.2f")
  println(f"Throughput:     $throughput%.2f M msg/sec")

  Await.result(system.whenTerminated, 10.seconds)
}
